### PR TITLE
Tests for new apis and code coverage 90%

### PR DESCRIPTION
**Pull request by harshithaputtaswamy**

### CODE REVIEW

File: apps/fyle/tasks.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def create_expense_groups(workspace_id: int, fund_source: List[str], task_log: T
     task_log.detail = {
         'message': 'Creating expense groups'
     }
-    task_log.save()
+    task_log.save()             #TODO: doubt
```
Comment (NileshPant1999) [inline apps/fyle/tasks.py:70]: what doubt you are having here

File: fyle_qbo_api/settings.py [hunk 2]
```diff
@@ -111,7 +111,8 @@
 Q_CLUSTER = {
     'name': 'fyle_quickbooks_api',
     'save_limit': 0,
-    'workers': os.environ.get('NO_WORKERS', 4),
+    # 'workers': os.environ.get('NO_WORKERS', 4),
```
Comment (NileshPant1999) [inline fyle_qbo_api/settings.py:114]: why are we changing this

File: tests/test_fyle/test_models.py [hunk 3]
```diff
@@ -9,15 +10,65 @@ def test_default_fields():
...
+    expense_group_settings.save()
+    
+    expense_groups = _group_expenses([], ['claim_number', 'fund_source', 'projects', 'employee_email', 'report_id'], 4)
+    assert expense_groups == []
+
+    ExpenseGroup.create_expense_groups_by_report_id_fund_source([expense_objects], workspace_id)
+
+    expense_groups = ExpenseGroup.objects.last()
+    assert expense_groups.exported_at == None
+
```
Comment (NileshPant1999) [inline tests/test_fyle/test_models.py:72]: can remove extra lines

File: tests/test_fyle/test_tasks.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    assert task_log.status == 'FAILED'
+
+    expense_group_settings = ExpenseGroupSettings.objects.get(workspace_id=1)
+    expense_group_settings.delete()
+
+    create_expense_groups(1, ['PERSONAL', 'CCC'], task_log)
+
+    task_log = TaskLog.objects.get(workspace_id=1)
+    assert task_log.status == 'FATAL' 
+
```
Comment (NileshPant1999) [inline tests/test_fyle/test_tasks.py:56]: can remove extra lines

File: tests/test_fyle/test_views.py [hunk 5]
```diff
@@ -284,11 +338,43 @@ def test_projects_view(api_client, test_connection):
...
+    url = reverse('sync-expense-groups', 
+        kwargs={
+                'workspace_id': 3
+            }
+        )
+
+    api_client.credentials(HTTP_AUTHORIZATION='Bearer {}'.format(access_token))
+
+    response = api_client.post(url)
+    assert response.status_code == 200
```
Comment (NileshPant1999) [inline tests/test_fyle/test_views.py:380]: new line

File: tests/test_mappings/fixtures.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    ],
+
+    "cost_center_payload": [
+        {
+            "name": "Accessories",
+            "is_enabled": True,
+            "description": "Cost Center - Accessories, Id - 10"
+        },
+    ]
+}
```
Comment (NileshPant1999) [inline tests/test_mappings/fixtures.py:13]: new line

File: tests/test_mappings/test_signals.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            'type': 'CATEGORY_MAPPING',
+            'error_title': category_attribute.value,
+            'error_detail': 'Category mapping is missing',
+            'is_resolved': False
+        }
+    )
+
+    mapping = Mapping(
+        source_type='CATEGORY',
+        destination_type='ACCOUNT',
```
Comment (NileshPant1999) [inline tests/test_mappings/test_signals.py:28]: why not add source value

File: tests/test_mappings/test_signals.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+       source_employee_id=2082,
+        workspace_id=2
+    )
+    employee_mapping.destination_employee_id = 748
+    employee_mapping.save()
+
+    error = Error.objects.filter(expense_attribute_id=employee_mapping.source_employee_id).first()
+
+    assert error.is_resolved == True
+
```
Comment (NileshPant1999) [inline tests/test_mappings/test_signals.py:68]: just two line gab will be enough

File: tests/test_workspaces/test_apis/test_export_settings/fixtures.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        'general_mappings': {'accounts_payable':{'name':'Accumulated Depreciation','id':'3'},'qbo_expense_account':{'name':'Auto','id':'95'},'bank_account':{'name':'','id':''},'default_ccc_account':{'name':'','id':''},'default_debit_card_account':{'name':'','id':''},'default_ccc_vendor':{'name':'Arun','id':'87'}
+        },
+        'workspace_id':9
+    },
+    'export_settings_missing_values': {
+        'expense_group_settings': {},
+        'workspace_general_settings': {},
+        'general_mappings': {}
+    },
+}
```
Comment (NileshPant1999) [inline tests/test_workspaces/test_apis/test_export_settings/fixtures.py:57]: new line

File: tests/test_workspaces/test_apis/test_import_settings/test_views.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    response = api_client.put(
+        url,
+        data=data['import_settings_missing_values'],
+        format='json'
+    )
+
+    assert response.status_code == 400
+
+# {'mapping_settings': [{'source_placeholder': [ErrorDetail(string='This field may not be blank.', code='blank')]}, {'source_placeholder': [ErrorDetail(string='This field may not be blank.', code='blank')]}, {'source_placeholder': [ErrorDetail(string='This field may not be blank.', code='blank')]}
```
Comment (NileshPant1999) [inline tests/test_workspaces/test_apis/test_import_settings/test_views.py:37]: why comment this

File: tests/test_workspaces/test_apis/test_map_employee/fixtures.py [hunk 11]
```diff
@@ -0,0 +1,6 @@
+data = {
+    'map_employees': {'workspace_general_settings': {'employee_field_mapping': 'EMPLOYEE', 'auto_map_employees': 'EMAIL'}},
+    'response': {"workspace_general_settings":{"employee_field_mapping":"EMPLOYEE","auto_map_employees":"EMAIL"},"workspace_id":8},
+    'invalid_auto_map_employees': {'workspace_general_settings': {'employee_field_mapping': 'EMPLOYEE', 'auto_map_employees': 'EMAILs'}},
+    'missing_employee_mapping': {'workspace_general_settings': {'employee_field_mapping': '', 'auto_map_employees': 'EMAILs'}},
+}
```
Comment (NileshPant1999) [inline tests/test_workspaces/test_apis/test_map_employee/fixtures.py:6]: new line

File: apps/fyle/tasks.py (hunk 1 above)
Comment (ashwin1111) [inline apps/fyle/tasks.py:70]: what's the doubt?

File: apps/quickbooks_online/utils.py [hunk 12]
```diff
@@ -1,9 +1,12 @@
 import base64
+from multiprocessing import connection
```
Comment (ashwin1111) [inline apps/quickbooks_online/utils.py:2]: remove unused import

File: apps/quickbooks_online/utils.py [hunk 13]
```diff
@@ -49,11 +52,18 @@ def __init__(self, credentials_object: QBOCredential, workspace_id: int):
         client_id = settings.QBO_CLIENT_ID
         client_secret = settings.QBO_CLIENT_SECRET
         environment = settings.QBO_ENVIRONMENT
+        refresh_token = ''
+
+        if 'QBO_REFRESH_TOKENS' in os.environ:
```
Comment (ashwin1111) [inline apps/quickbooks_online/utils.py:57]: rename to `QBO_TESTS_REFRESH_TOKENS`
`if os.environ.get('QBO_REFRESH_TOKENS'):`

File: apps/quickbooks_online/utils.py [hunk 14]
```diff
@@ -63,6 +73,12 @@ def __init__(self, credentials_object: QBOCredential, workspace_id: int):
         credentials_object.refresh_token = self.connection.refresh_token
         credentials_object.save()
 
+        if 'QBO_REFRESH_TOKENS' in os.environ:
```
Comment (ashwin1111) [inline apps/quickbooks_online/utils.py:76]: `if os.environ.get('QBO_REFRESH_TOKENS'):`

File: apps/quickbooks_online/views.py [hunk 15]
```diff
@@ -586,7 +586,7 @@ def post(self, request, *args, **kwargs):
         expense_group = ExpenseGroup.objects.get(pk=expense_group_id)
         task_log = TaskLog.objects.get(pk=task_log_id)
 
-        create_cheque(expense_group, task_log)
+        create_cheque(expense_group, task_log_id)
```
Comment (ashwin1111) [inline apps/quickbooks_online/views.py:589]: this is fine, unused endpoint.

**Pull request merged**